### PR TITLE
setting sensible default based on exact angular integration

### PR DESCRIPTION
**Pull request by camelto2**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

here is a solution to the reasonable default choice of angular quadrature choice based on the value of lmax in the pseudo. The choices of Nrule are based on [J. Chem. Phys. 95 (3467) (1991)](https://aip.scitation.org/doi/10.1063/1.460849)
I'm setting this as a WIP until these defaults can be validated, but I wanted to present a simple solution. 

Assuming the defaults are reasonable in most cases, this should be fine. If the user wants to override this, adding "nrule=XXX" to the pseudopotential xml file is all that is needed since it is already parsed. I think this is the simplest solution and requires the least changes to the code.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
My workstation

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
See also https://github.com/QMCPACK/qmcpack/issues/557

File: src/QMCHamiltonians/ECPComponentBuilder.cpp [hunk 1]
```diff
@@ -207,6 +207,43 @@ bool ECPComponentBuilder::put(xmlNodePtr cur)
   }
   if (pp_nonloc)
   {
+    if (Nrule == -1)
+    {
+      app_log() << "Quadrature rule not determined from pseudopotential xml file. Setting sensible default" << std::endl;
```
Comment (prckent) [inline src/QMCHamiltonians/ECPComponentBuilder.cpp:212]: Move after the switch and print which nrule was chosen.

Comment (ye-luo) [reply src/QMCHamiltonians/ECPComponentBuilder.cpp:212]: Please move the whole switch to a function.

Comment (prckent) [reply src/QMCHamiltonians/ECPComponentBuilder.cpp:212]: @ye-luo I think this is fine as is. Function won't help much, just more indirection for the reader. A strong argument for a function would be needing this switch more than once. However this is init only and we don't.

Comment (camelto2) [reply src/QMCHamiltonians/ECPComponentBuilder.cpp]: I moved the print statement. I can change it to a function if necessary

**Comment by camelto2** [comment]
I think leaving the override for the Nrule in the pseudopotential xml file is simplest solution.

However, if you want the user override to be at the qmcpack input level, then it can be added in the input like
`
            <pseudo elementType="W" href="W.qmcpack.xml" nrule="X"/>
`
 and the nrule could be passed as a constructor argument to ECPComponentBuilder. Then if it doesn't see it as an argument, it would set the defaults as above. 
 
 That approach would potentially be problematic though. Nrule is already read from the pseudo input file. So to avoid having overrides in two different places that potentially could disagree with one another, we would need to remove the Nrule setting from the pseudo xml file. That would change the expected behavior of the code for users who have been setting Nrule already in the pseudopotential file.

File: src/QMCHamiltonians/ECPComponentBuilder.cpp [hunk 2]
```diff
@@ -207,6 +207,43 @@ bool ECPComponentBuilder::put(xmlNodePtr cur)
...
+      //nonlocal channels.
+      //From J. Chem. Phys. 95 (3467) (1991)
+      switch (pp_nonloc->lmax)
+      {
+      case 0:
+        Nrule = 1;
+        break;
+      case 1:
+      case 2:
+        Nrule = 2;
```
Comment (jtkrogel) [inline src/QMCHamiltonians/ECPComponentBuilder.cpp:222]: I recommend we keep a floor at nrule=4.  I'm pretty sure some d-local potentials have higher variance at these lower settings.

**Comment by prckent** [comment]
Print a warning if nlrule is already set. Use case is someone testing the required integration rule. They should not have to modify the ECP file to do that. No need to over think this.

File: src/QMCHamiltonians/ECPComponentBuilder.cpp (hunk 2 above)
Comment (jtkrogel) [reply src/QMCHamiltonians/ECPComponentBuilder.cpp:222]: Also, we don't want to degrade the statistical performance of existing inputs.

**Comment by prckent** [comment]
I am also OK with leaving the default minimum at nrule=4 as Jaron suggests. Changing this might surprise a few people. But we allow override.


( These options become more important the slower the basis function and single orbital evaluation is. Not very important for splines, but potentially important for other choices.)

File: src/QMCHamiltonians/ECPComponentBuilder.cpp (hunk 2 above)
Comment (camelto2) [reply src/QMCHamiltonians/ECPComponentBuilder.cpp]: changed the floor to nrule=4

**Comment by camelto2** [comment]
I changed the nrule=4 as Jaron suggested. 
I also added the ability to modify nrule in the qmcpack input. 
Essentially, we only need to add 
<pseudo elementType="" href="" nrule=""/> to override the default settings. 

In the case that nrule is set in the pseudopotential input file, it prints a warning that it should be set/modified from the qmcpack input file, but proceeds using the value set in the pseudopotential file.

**Comment by ye-luo** [comment]
> In the case that nrule is set in the pseudopotential input file, it prints a warning that it should be set/modified from the qmcpack input file, but proceeds using the value set in the pseudopotential file.

Not getting the logic here. I was thinking of if nrule
1. not in qmcpack input but in pp file. take pp file value. no warning
2. neither in qmcpck input nor in pp file, use qmcpack fallback default which needs to be mentioned as a warning.
3. in qmcpack but not in pp file. take qmcpack input. 
4. both qmcpack and pp file provide nrule. Print a warning that qmcpack input overrides the value.

always print warning that pp file doesn't have nrule.

**Comment by camelto2** [comment]
> > In the case that nrule is set in the pseudopotential input file, it prints a warning that it should be set/modified from the qmcpack input file, but proceeds using the value set in the pseudopotential file.
> 
> Not getting the logic here. I was thinking of if nrule
> 
> 1. not in qmcpack input but in pp file. take pp file value. no warning
> 2. neither in qmcpck input nor in pp file, use qmcpack fallback default which needs to be mentioned as a warning.
> 3. in qmcpack but not in pp file. take qmcpack input.
> 4. both qmcpack and pp file provide nrule. Print a warning that qmcpack input overrides the value.
> 
> always print warning that pp file doesn't have nrule.

In almost all the pp files I've found, (pseudopotentiallibrary, pseudos in test directory, etc) the xml file does not have nrule by default. So it seems that almost every run will have a warning printed since almost no pseudo sets nrule. 

1, 2 & 3 are currently satisfied.  I can modify it to let qmcpack override instead of the pseudo input. The only change needed is to make the qmcpack input > pseudo input and decide where exactly we want warning vs. just printing the information.

**Comment by camelto2** [comment]
Ok, warnings should be consistent with @ye-luo's previous comment. I also moved the default setting to the same location inside ECPComponentBuilder::buildSemiLocalAndLocal so that it is all in the same place

**Comment by prckent** [comment]
Echoing my agreement with Ye's list of warnings priorities and overrides.

Indeed, many existing runs and potentials will result in warnings being printed, but I think this the appropriate.

File: src/QMCHamiltonians/ECPComponentBuilder.2.cpp [hunk 3]
```diff
@@ -49,17 +49,66 @@ void ECPComponentBuilder::buildSemiLocalAndLocal(std::vector<xmlNodePtr>& semiPt
...
+  aAttrib.add(quadRule, "nrule");
   aAttrib.add(Srule, "srule");
   aAttrib.add(nso, "npots-so");
 
   xmlNodePtr cur_semilocal = semiPtr[0];
   aAttrib.put(cur_semilocal);
+
+  if (quadRule > -1 && Nrule > -1)
+  {
+    app_warning() << " Nrule setting found in both qmcpack input and pseudopotential file."
```
Comment (ye-luo) [inline src/QMCHamiltonians/ECPComponentBuilder.2.cpp:67]: Please echo both values.

File: src/QMCHamiltonians/ECPComponentBuilder.2.cpp [hunk 4]
```diff
@@ -49,17 +49,66 @@ void ECPComponentBuilder::buildSemiLocalAndLocal(std::vector<xmlNodePtr>& semiPt
...
+                  << " Using nrule setting in qmcpack input file\n";
+  }
+  else if (quadRule > -1 && Nrule == -1)
+  {
+    app_log() << " Nrule setting found in pseudopotential file.\n";
```
Comment (ye-luo) [inline src/QMCHamiltonians/ECPComponentBuilder.2.cpp:72]: "Nrule setting found in pseudopotential file and used."
std::endl incluses a flush. it is preferred to "\n"

File: src/QMCHamiltonians/ECPComponentBuilder.2.cpp [hunk 5]
```diff
@@ -49,17 +49,66 @@ void ECPComponentBuilder::buildSemiLocalAndLocal(std::vector<xmlNodePtr>& semiPt
...
+    Nrule = quadRule;
+  }
+  else if (quadRule == -1 && Nrule > -1)
+    app_log() << " Nrule setting found in qmcpack input file.\n";
```
Comment (ye-luo) [inline src/QMCHamiltonians/ECPComponentBuilder.2.cpp:76]: as above.

File: src/QMCHamiltonians/ECPComponentBuilder.2.cpp [hunk 6]
```diff
@@ -49,17 +49,66 @@ void ECPComponentBuilder::buildSemiLocalAndLocal(std::vector<xmlNodePtr>& semiPt
...
+  
+  if (Nrule == -1)
```
Comment (ye-luo) [inline src/QMCHamiltonians/ECPComponentBuilder.2.cpp:78]: Prefer this to be an else connected tot he above to make it clear 4 cases.

**Review by ye-luo — COMMENTED** [review]
Only some minor issue left. Thanks.

**Comment by ye-luo** [comment]
Still WIP?

**Comment by ye-luo** [comment]
> Echoing my agreement with Ye's list of warnings priorities and overrides.
> 
> Indeed, many existing runs and potentials will result in warnings being printed, but I think this the appropriate.

In LLVM, to differentiate from Warning, Remark is used.

**Comment by camelto2** [comment]
> Still WIP?

Nope, the WIP was originally for if we were going to change to smaller nrules as the default for smaller angular momentum channel projector. But with the default set to nrule=4 for most pseudos, I think we are good to go.

**Comment by ye-luo** [comment]
Okay to test

**Review by ye-luo — APPROVED** [review]

**Pull request merged**